filebeat/input/log/input.go

```diff
@@ -87,11 +88,21 @@ func NewInput(
 	if err != nil {
 		return nil, err
 	}
+	defer func() {
+		if cleanupNeeded {
+			out.Close()
+		}
+	}()
 
 	// stateOut will only be unblocked if the beat is shut down.
 	// otherwise it can block on a full publisher pipeline, so state updates
 	// can be forwarded correctly to the registrar.
 	stateOut := channel.CloseOnSignal(channel.SubOutlet(out), context.BeatDone)
+	defer func() {
+		if cleanupNeeded {
+			stateOut.Close()
+		}
+	}()
```

Comment:
In go-txfile we have a package for this :)
https://github.com/elastic/go-txfile/blob/master/internal/cleanup/cleanup.go

it's used like this:

```
ok := false
defer cleanup.IfNot(&ok, func() { 
  ...
})

...

ok = true
return // something
```

Reply:
I have reordered the code so this is not needed, tell me what you think. The only risk is that the harvester created to check the config will have nil outleters, but they shouldn't be used on creation.

Reply:
@urso `cleanup` is under `internal` and then it cannot be imported from beats, I have added a similar convenience helper.